core/src/main/java/hivemall/mf/CofactorModel.java

```diff
@@ -0,0 +1,629 @@
+    }
+
+    /**
+     * Calculate W' x W plus regularization matrix
+     */
+    protected static RealMatrix calculateWTWpR(Map<String, RealVector> W, int numFactors, float c0, RealMatrix idMatrix, float lambda) {
+        RealMatrix WTW = calculateWTW(W, numFactors, c0);
+        RealMatrix R = calculateR(idMatrix, lambda, numFactors);
+        return WTW.add(R);
+    }
+
+    protected static RealMatrix calculateR(RealMatrix idMatrix, float lambda, int numFactors) {
+        if (idMatrix == null) {
+            idMatrix = new Array2DRowRealMatrix(MatrixUtils.eye(numFactors));
+        }
+        return idMatrix.scalarMultiply(lambda);
+    }
+
+    protected static List<Feature> filterTrainableFeatures(Feature[] features, Map<String, RealVector> weights) {
+        List<Feature> trainableFeatures = new ArrayList<>();
+        String fName;
+        for (Feature f : features) {
+            fName = f.getFeature();
+            if (isTrainable(fName, weights)) {
+                trainableFeatures.add(f);
+            }
+        }
+        return trainableFeatures;
+    }
+
+    protected static RealVector solve(RealMatrix B, RealVector a) {
+        // b * x = a
+        // solves for x
+        SingularValueDecomposition svd = new SingularValueDecomposition(B);
+        return svd.getSolver().solve(a);
+
+    }
+
+    protected static RealMatrix calculateWTW(Map<String, RealVector> weights, int numFactors, float constant) {
+        RealMatrix WTW = new Array2DRowRealMatrix(numFactors, numFactors);
+        int i = 0, j = 0;
+        for (int f = 0; f < numFactors; f++) {
+            for (int ff = 0; ff < numFactors; ff++) {
+                double val = constant * dotFactorsAlongDims(weights, f, ff);
+                WTW.setEntry(f, ff, val);
+            }
+        }
+        return WTW;
+    }
+
+    protected static RealMatrix calculateWTWSubset(List<Feature> subset, Map<String, RealVector> weights, int numFactors, float constant) {
+        // equivalent to `B_u.T.dot((c1 - c0) * B_u)` in cofacto.py
+        RealMatrix delta = new Array2DRowRealMatrix(numFactors, numFactors);
+        int i = 0, j = 0;
+        for (int f = 0; f < numFactors; f++) {
+            for (int ff = 0; ff < numFactors; ff++) {
+                double val = constant * dotFactorsAlongDims(subset, weights, f, ff);
+                delta.setEntry(f, ff, val);
+            }
+        }
+        return delta;
+    }
+
+    private static double dotFactorsAlongDims(List<Feature> keys, Map<String, RealVector> weights, int dim1, int dim2) {
+        double result = 0.d;
+        for (Feature f : keys) {
+            RealVector vec = getFactorVector(f.getFeature(), weights);
+            result += vec.getEntry(dim1) * vec.getEntry(dim2);
+        }
+        return result;
+    }
+
+    private static double dotFactorsAlongDims(Map<String, RealVector> weights, int dim1, int dim2) {
+        double result = 0.d;
+        for (String key : weights.keySet()) {
```

Comment:
Why not simply 

```
        for(RealVector vec : weights.values()) {
           ...      
        }
```

What's expected behavior when `vec.getEntry(dim1/2)` returns null?

Reply:
`RealVector` implementation has now changed to `double[]`, so by default it should be `0`. In any case, all entries in `weights` have been initialized to uniform random or gaussian so the array entries should be valid.